src/libraries/Native/Unix/System.Globalization.Native/pal_icushim.c

```diff
@@ -353,23 +395,50 @@ int32_t GlobalizationNative_LoadICU()
             return FALSE;
         }
     }
-#endif
+#endif // TARGET_WINDOWS || TARGET_OSX
 
-    // Get pointers to all the ICU functions that are needed
-#define PER_FUNCTION_BLOCK(fn, lib) \
-    c_static_assert_msg((sizeof(#fn) + MaxICUVersionStringLength + 1) <= sizeof(symbolName), "The symbolName is too small for symbol " #fn); \
-    sprintf(symbolName, #fn "%s", symbolVersion); \
-    fn##_ptr = (__typeof(fn)*)dlsym(lib, symbolName); \
-    if (fn##_ptr == NULL) { fprintf(stderr, "Cannot get symbol %s from " #lib "\nError: %s\n", symbolName, dlerror()); abort(); }
+    FOR_ALL_ICU_FUNCTIONS
+    return TRUE;
+}
+
+void GlobalizationNative_InitICUFunctions(void* icuuc, void* icuin, const char* version, const char* suffix)
+{
+    assert(icuuc != NULL);
+    assert(icuin != NULL);
+    assert(version != NULL);
+    
+    libicuuc = icuuc;
+    libicui18n = icuin;
+    int major = -1;
+    int minor = -1;
+    int build = -1;
+
+    char symbolName[SYMBOL_NAME_SIZE];
+    char symbolVersion[MaxICUVersionStringWithSuffixLength + 1]="";
+    char symbolSuffix[SYMBOL_CUSTOM_SUFFIX_SIZE]="";
 
+    sscanf(version, "%d.%d.%d", &major, &minor, &build);
+
+    if (suffix != NULL)
+    {
+#if defined(TARGET_WINDOWS)
+    sprintf_s(symbolSuffix, SYMBOL_CUSTOM_SUFFIX_SIZE, "_%s", suffix);
+#else
+    sprintf(symbolSuffix, "_%s", suffix);
```

Comment:
We have a sprintf_s polyfill in the coreclr PAL:
https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/coreclr/src/pal/inc/mbusafecrt.h#L87
Have we considered using something like that here?  Or alternatively manually validate that strlen(suffix) + 1 <= SYMBOL_CUSTOM_SUFFX_SIZE?

Reply:
System.Globalization.Native library lives in libraries because it is used by both mono and coreclr, so I think it will bring much more complex build infra to include the coreclr headers and we would need to add the equivalent headers to mono once System.Globalization.Native is linked into the mono runtime.

I think I'll add validation instead.